File: Basic/src/main/java/org/myspecialway/Alarm.java
package org.myspecialway;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

public class Alarm {
    AlarmManager am;
    Context context;
    PendingIntent pendingIntent;
    private static Alarm alarmInstance = null;

    public static Alarm getInstance(Context cont){
        if (alarmInstance == null){
            alarmInstance = new Alarm(cont);
        }
        return alarmInstance;
    }
    private Alarm(Context cont){
        context = cont;
        am = (AlarmManager)context.getSystemService(Context.ALARM_SERVICE);
    }


    /**
     * Schedule a new alarm x seconds from now.
     * @param secondsFromNow
     */
    public void scheduleAlarm(int secondsFromNow){
        Intent intent = new Intent(context, AlarmReceiver.class);
        pendingIntent = PendingIntent.getBroadcast(context, 0, intent, 0);

        long triggerAtMillis = System.currentTimeMillis() + (secondsFromNow * 1000);
        Log.d("Peleg", String.valueOf(triggerAtMillis));
        am.setExact(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent);
        Log.d("Peleg", "Alarm scheduled");
    }

    /**
     * Cancel an existing alarm, if any
     */
    public void cancelAlarm(){
        if(pendingIntent != null) {
            am.cancel(pendingIntent);
        }
    }

    public static class AlarmReceiver extends BroadcastReceiver {
        public AlarmReceiver(){

        }

        @Override
        public void onReceive(Context context, Intent intent) {
            Log.d("Peleg", "Alarm fired");
            Notifications nm = new Notifications(context);
            ClassDetails cd = new ClassDetails("32","34","מדעים", "1");
            nm.showNavigationNotification(cd);
        }
    }


}

File: app/src/main/java/org/myspecialway/notifications/Alarm.kt
package org.myspecialway.notifications

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import org.myspecialway.common.Navigation
import org.myspecialway.ui.agenda.ScheduleRenderModel

class Alarm(private val context: Context) {
    private var alarmManager: AlarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager

    private val alarmsQueue = mutableListOf<PendingIntent>()

    fun scheduleAlarm(scheduleModel: ScheduleRenderModel) {
        alarmsQueue.cancelAll()

        val intent = Intent(context, AlarmReceiver::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        intent.putExtra(NotificationActivity.NOTIFICATION_TITLE, scheduleModel.title)
        val id = System.currentTimeMillis().toInt()
        val pendingIntent = PendingIntent.getBroadcast(context, id, intent, PendingIntent.FLAG_ONE_SHOT)
        alarmsQueue.add(pendingIntent)
        val triggerAtMillis = scheduleModel.time!!.date.time - System.currentTimeMillis()
        alarmManager.set(
                AlarmManager.RTC_WAKEUP,
                System.currentTimeMillis() + triggerAtMillis,
                pendingIntent
        )
    }

    class AlarmReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val scheduleTitle = intent.getStringExtra(NotificationActivity.NOTIFICATION_TITLE)
            Navigation.toNotificationActivity(context, ScheduleRenderModel(title = scheduleTitle))
        }
    }
}

fun MutableList<PendingIntent>.cancelAll() {
    clear()
    forEach {it.cancel() }
}